include/cpp2util.h

```diff
@@ -1269,132 +1345,197 @@ inline constexpr auto is( X const& x, bool (*value)(X const&) ) -> bool {
+
+template< typename X, typename C >
+concept brace_initializable_to = requires (X x) { C{x}; };
+
+template< typename From, typename To > // https://eel.is/c++draft/dcl.init.list#7
+concept narrowing_to = !brace_initializable_to<std::remove_cvref_t<From>, To>;
+    // // [dcl.init.list] 7.1
+    // (std::floating_point<From> && std::integral<To>) ||
+    // // [dcl.init.list] 7.2
+    // (std::floating_point<From> && std::floating_point<To> && type_bigger_than<From, To>) ||
+    // // [dcl.init.list] 7.3
+    // (std::integral<From> && std::floating_point<To>) ||
+    // (enum_type<From> && std::floating_point<To>) ||
+    // // [dcl.init.list] 7.4
+    // (std::integral<From> && std::integral<To> && type_bigger_than<From, To>) ||
+    // (enum_type<From> && std::integral<To> && type_bigger_than<From, To>) ||
+    // // [dcl.init.list] 7.5
+    // (pointer<From> && std::same_as<To, bool>);
 
 template <typename... Ts>
 inline constexpr auto program_violates_type_safety_guarantee = sizeof...(Ts) < 0;
 
-//  For literals we can check for safe 'narrowing' at a compile time (e.g., 1 as std::size_t)
 template< typename C, auto x >
-concept castable = requires{ C{x}; } || (
-    std::is_integral_v<C> &&
-    std::is_integral_v<CPP2_TYPEOF(x)> &&
-    !(static_cast<CPP2_TYPEOF(x)>(static_cast<C>(x)) != x ||
-        (
-            (std::is_signed_v<C> != std::is_signed_v<CPP2_TYPEOF(x)>) &&
-            ((static_cast<C>(x) < C{}) != (x < CPP2_TYPEOF(x){}))
-        )
-    )
-);
+concept castable = requires{ C{x}; };
 
 //  As
 //
 
-template< typename C, typename X>
-auto as( X const& ) -> auto {
-    return nonesuch;
+template <typename C,        typename X > auto as( X&& ) -> nonesuch_<> { 
+        //std::cout << __PRETTY_FUNCTION__ << " " << __LINE__ << std::endl;
+    return {}; }
+template< arithmetic C, arithmetic X > requires narrowing_to<X, C>
+auto as( X&& ) -> nonesuch_<casting_errors::narrowing_cast> {
+        //std::cout << __PRETTY_FUNCTION__ << " " << __LINE__ << std::endl;
+     return {}; }
+
+template< arithmetic C, arithmetic X >
+    requires (!narrowing_to<std::remove_cvref_t<X>,C>)
+           && (!std::same_as<std::remove_cvref_t<X>,C>)
+auto as( X&& x ) -> decltype(auto) {
+        //std::cout << __PRETTY_FUNCTION__ << " " << __LINE__ << std::endl;
+    return forward_like<forward_like_t<C,X&&>>(x);
+}
+
+template< arithmetic C, typename X >
+    requires arithmetic<typename std::remove_cvref_t<X>::value_type>
+          && (!narrowing_to<typename std::remove_cvref_t<X>::value_type,C>) 
+          && (!std::same_as<std::remove_cvref_t<X>,C>)
+          && (!can_bound_to<typename std::remove_cvref_t<X>::value_type,C>)
+auto as( X&& x ) -> decltype(auto) {
+        //std::cout << __PRETTY_FUNCTION__ << " " << __LINE__ << std::endl;
+    return forward_like<forward_like_t<C,X&&>>(*x);
+}
+
+template <typename T>
+using pointee_t = std::remove_cvref_t<decltype(*std::declval<std::remove_cvref_t<T>>())>;
```

Comment:
Seems like you want `std::iter_value_t<T>`.

Reply:
`pointee_t` is also used for things like `std::optional`, smart pointers, etc... in general, anything you can dereference.

Reply:
It works with those: <https://compiler-explorer.com/z/af1c11aa5>.